### PR TITLE
Improve inifile handling.

### PR DESCRIPTION
**Pull request by pgScorpio**
Improved inifile handling which allows saving inifiles with sections for client and server and including fileversion info.

**Short description of changes**

Get/Set iniset functions now take a document node parameter instead of document and section name parameters which is more efficient and now allows using sections (section name in the current implementation does not work as expected. Since they only work if all section names are the same.)
client and server can now use the same inifile.
Even using subsections in the inifile for i.e. stored fadersettings and/or soundcard settings will be possible too now.

Also added Get/Set iniset functions for base64 encoded values.

HEADLESS server can now use an inifile, but will only do so if a --inifile parameter is given on the commandline.

Also adding a "fileversion" setting in the new inifile which will make backwards compatibility handling easier on future changes.

NOTE: Old "single section" inifiles can still be read but files will be saved in the new "multi section" format. If we still want to use the current "single section" format this is still possible by changing some defines. (See settings.h)

CHANGELOG: 
changed inifile handling, headless server can now use inifile too.

**Context: Fixes an issue?**
related to #555 #1973
Together with my messages and commandline parameter PR's this is a good base for a complete settings handling overhaul.

**Does this change need documentation? What needs to be documented and how?**
Since a headless server now CAN use a inifile, probably yes.

**Status of this Pull Request**
ready to merge

**What is missing until this pull request can be merged?**
review and successful autobuilds.

Tested on Windows client and server, so should work on other OS'es too.
## Checklist

-  [x] I've verified that this Pull Request follows the [general code principles](https://github.com/jamulussoftware/jamulus/blob/master/CONTRIBUTING.md#jamulus-projectsource-code-general-principles)
-  [x] I tested my code and it does what I want
-  [x] My code follows the [style guide](https://github.com/jamulussoftware/jamulus/blob/master/CONTRIBUTING.md#source-code-consistency) 
-  [x] I waited some time after this Pull Request was opened and all GitHub checks completed without errors. 
-  [x] I've filled all the content above

### CODE REVIEW

**Comment by ann0see** [comment]
I'd like to leave this PR up to Tony and Peter. I hope you understand. 

- We really need to get your PRs in -

**Comment by pgScorpio** [comment]
@ann0see 

> I'd like to leave this PR up to Tony and Peter. I hope you understand.

I would understand if they had reacted on my requests to cooperate on this issue... ;=)
But I'm always open for further discussion...

> _We really need to get your PRs in some day soon_

Yes it's about time, but reviews seem to be the problem, my oldest PR is already waiting over a month now !
If this doesn't change it is of no use that I continue with any redesigns....
You asked me to split the redesigns in multiple PS's, but if every PR is going to take this long to merge, the total completion would take me far too long...

**Comment by ann0see** [comment]
@henkdegroot not sure if you can help reviewing this PR?

File: src/main.cpp [hunk 1]
```diff
@@ -797,7 +797,7 @@ int main ( int argc, char** argv )
     // be located in the install directory of the software by the installer.
     // Here, we set the path to our application path.
     QDir ApplDir ( QApplication::applicationDirPath() );
-    pApp->addLibraryPath ( QString ( ApplDir.absolutePath() ) );
+    pAPPLICATION->addLibraryPath ( QString ( ApplDir.absolutePath() ) );
```
Comment (ann0see) [inline src/main.cpp:800]: Why this change? pApp is short and concise?

**Review by ann0see — COMMENTED** [review]
@atsampson can you have a quick look at this PR? You seem to have more C++ knowledge?

File: src/main.cpp (hunk 1 above)
Comment (pgScorpio) [reply src/main.cpp:800]: > Why this change? pApp is short and concise?

In this way we are shure to call with a pointer of the correct type.
Note that this is not always the case in the current code. And though QApplication is a descendant of QCoreApplication, and has a lot of the same functions, many of these functions are NOT virtual.
This might be the cause of some GUI problems with android and iOS, since calling via a QCoreApplication pointer can skip some GUI initialization.

**Comment by henkdegroot** [comment]
> @henkdegroot not sure if you can help reviewing this PR?

Not really comfortable to comment/review this. I can run a test to check if the app still works okay.

**Comment by ann0see** [comment]
Yeah. That's still a good value ;-)

File: src/settings.h [hunk 2]
```diff
@@ -38,37 +38,91 @@
 #include "server.h"
 #include "util.h"
 
+/* Defines ********************************************************************/
+
+// In the new inifile format, both server and client settings can be stored in the same inifile.
```
Comment (ann0see) [inline src/settings.h:43]: I wouldn't call it "new" format. Better give a Jamulus version here.

Comment (pgScorpio) [reply src/settings.h:43]: > I wouldn't call it "new" format. Better give a Jamulus version here.

Why isn't it "new" ??? Did you read the next comment lines too?? If we save in the "new" format it can no longer be read by "old" versions of Jamulus...
And I can't give a Jamulus version yet, since I don't know when it will be implemented...

**Comment by ann0see** [comment]
@softins once you have more time for Jamulus could you please review this PR and voice your opinion

**Comment by ann0see** [comment]
Closing as no review for long time. If there's interest, please re-open.
